File: leetcode/83.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def remove_duplicates(node: Optional[ListNode]) -> None:
            if node == None:
                return

            if node.next  == None:
                return

            if node.val == node.next.val:
                node.next = node.next.next
                return remove_duplicates(node)

            return remove_duplicates(node.next)

        remove_duplicates(head)

        return head

File: leetcode/83.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteDuplicates = function(head) {
  const dedup = node => {
    if (node === null) return;
    if (node.next === null) return;

    if (node.val === node.next.val) {
      // remove next
      //console.log("removing:", node.next.val);
      node.next = node.next.next;
      dedup(node);
    } else {
      dedup(node.next);
    }
  };

  dedup(head);

  return head;
};
